selfdrive/ui/sound.cc

```diff
@@ -1,91 +1,143 @@
-  { cereal::CarControl::HUDControl::AudibleAlert::CHIME_WARNING2, "../assets/sounds/warning_2.wav", false },
-  { cereal::CarControl::HUDControl::AudibleAlert::CHIME_WARNING2_REPEAT, "../assets/sounds/warning_2.wav", true },
-  { cereal::CarControl::HUDControl::AudibleAlert::CHIME_WARNING_REPEAT, "../assets/sounds/warning_repeat.wav", true },
-  { cereal::CarControl::HUDControl::AudibleAlert::CHIME_ERROR, "../assets/sounds/error.wav", false },
-  { cereal::CarControl::HUDControl::AudibleAlert::CHIME_PROMPT, "../assets/sounds/error.wav", false },
-  { cereal::CarControl::HUDControl::AudibleAlert::NONE, NULL, false },
-};
+  setVolume(volume);
+  return true;
+}
 
-sound_file* get_sound_file(AudibleAlert alert) {
-  for (sound_file *s = sound_table; s->alert != cereal::CarControl::HUDControl::AudibleAlert::NONE; s++) {
-    if (s->alert == alert) {
-      return s;
+AudibleAlert Sound::currentPlaying() {
+  if (currentSound_ != AudibleAlert::NONE) {
+    auto playItf = player_.at(currentSound_)->playItf;
+    SLuint32 state;
+    if (SL_RESULT_SUCCESS == (*playItf)->GetPlayState(playItf, &state) &&
+        (state == SL_PLAYSTATE_STOPPED || state == SL_PLAYSTATE_PAUSED)) {
+      currentSound_ = AudibleAlert::NONE;
     }
   }
-
-  return NULL;
+  return currentSound_;
 }
 
-void play_alert_sound(AudibleAlert alert) {
-  sound_file* sound = get_sound_file(alert);
-  char* error = NULL;
-
-  slplay_play(sound->uri, sound->loop, &error);
-  if(error) {
-    LOGW("error playing sound: %s", error);
+void SLAPIENTRY slplay_callback(SLPlayItf playItf, void *context, SLuint32 event) {
+  Sound::Player *s = reinterpret_cast<Sound::Player *>(context);
+  if (event == SL_PLAYEVENT_HEADATEND && s->repeat > 1) {
+    --s->repeat;
+    (*playItf)->SetPlayState(playItf, SL_PLAYSTATE_STOPPED);
+    (*playItf)->SetMarkerPosition(playItf, 0);
+    (*playItf)->SetPlayState(playItf, SL_PLAYSTATE_PLAYING);
   }
 }
 
-void stop_alert_sound(AudibleAlert alert) {
-  sound_file* sound = get_sound_file(alert);
-  char* error = NULL;
-
-  slplay_stop_uri(sound->uri, &error);
-  if(error) {
-    LOGW("error stopping sound: %s", error);
+bool Sound::play(AudibleAlert alert, int repeat) {
+  if (currentSound_ != AudibleAlert::NONE) {
+    stop();
+  }
+  auto player = player_.at(alert);
+  SLPlayItf playItf = player->playItf;
+  player->repeat = repeat;
+  if (player->repeat > 0) {
+    ReturnOnError((*playItf)->RegisterCallback(playItf, slplay_callback, player), "Failed to register callback");
+    ReturnOnError((*playItf)->SetCallbackEventsMask(playItf, SL_PLAYEVENT_HEADATEND), "Failed to set callback event mask");
   }
-}
 
-void ui_sound_init() {
-  char *error = NULL;
-  slplay_setup(&error);
-  if (error) goto fail;
+  // Sets the player's volume
+  if (player->volume != volume_) {
+    SLmillibel vol = (volume_ < 0.01F ? -96.0F : 20 * log10(volume_)) * 100;
+    LogOnError((*(player->volumeItf))->SetVolumeLevel(player->volumeItf, vol), "Failed to set player volume");
```

Comment:
This works on my LEON.i have no other device to test it.

Reply:
We did testing with both devices at various volume levels to come up with the values, so I'd prefer to just leave the volume control the way it is.